metagraph/src/graph/graph_extensions/node_first_cache.hpp

```diff
@@ -0,0 +1,45 @@
+#ifndef __NODE_FIRST_CACHE_HPP__
+#define __NODE_FIRST_CACHE_HPP__
+
+#include <cache.hpp>
+#include <lru_cache_policy.hpp>
+
+#include "graph/representation/succinct/dbg_succinct.hpp"
+
+
+namespace mtg {
+namespace graph {
+
+class NodeFirstCache : public SequenceGraph::GraphExtension {
+  public:
+    using node_index = typename SequenceGraph::node_index;
+    using IncomingEdgeCallback = DeBruijnGraph::IncomingEdgeCallback;
+
+    NodeFirstCache() : dbg_succ_(nullptr), first_cache_(1) {}
+    NodeFirstCache(const DBGSuccinct &graph, size_t cache_size = 100'000)
+          : dbg_succ_(&graph), first_cache_(cache_size) {}
+
+    char get_first_char(node_index node) const;
+
+    void call_incoming_kmers(node_index node, const IncomingEdgeCallback &callback) const;
+
+    bool load(const std::string &) { throw std::runtime_error("Not implemented"); }
+    void serialize(const std::string &) const { throw std::runtime_error("Not implemented"); }
+
+    bool is_compatible(const SequenceGraph &graph, bool verbose = true) const;
+
+  private:
+    const DBGSuccinct *dbg_succ_;
+
+    using edge_index = boss::BOSS::edge_index;
+    mutable caches::fixed_sized_cache<edge_index, std::pair<edge_index, edge_index>,
+                                      caches::LRUCachePolicy<node_index>> first_cache_;
```

Comment:
Having a comment explaining what's in the cache would be nice

Reply:
done